File: boj_10974_모든 순열.cpp
//https://www.acmicpc.net/problem/10974
#include <iostream>
#include <vector>

using namespace std;

int n;
bool visited[9];

void permutation(int cnt,vector<int> result) {
	if (cnt == n) {
		for (int i = 0; i < cnt; i++) {
			cout << result[i] << ' ';
		}
		cout << '\n';
		return;
	}
	for (int i = 1; i <= n; i++) {
		if (visited[i]) continue;
		visited[i] = true;
		result[cnt] = i;
		permutation(cnt + 1,result);
		visited[i] = false;

	}
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
	cin >> n;

	vector<int> result(n);
	permutation(0,result);


	return 0;
}

File: 순열/boj_10974_모든 순열.kt
//https://www.acmicpc.net/problem/10974
import java.io.BufferedWriter

val br = System.`in`.bufferedReader()
fun getInt() = br.readLine().trim().toInt()

lateinit var visited: BooleanArray

fun permutation(len: Int, n: Int, result: IntArray, bw: BufferedWriter) {
    if (len == n) {
        bw.write("${result.filter { it != 0 }.joinToString(" ")}\n")
        return
    }
    for (i in 1..n) {
        if (visited[i]) continue
        result[len] = i
        visited[i] = true
        permutation(len+1, n, result, bw)
        result[len] = 0
        visited[i] = false
    }
}

fun main() {
    val n = getInt()
    val bw = System.out.bufferedWriter()
    visited = BooleanArray(n + 1)
    permutation(0, n, IntArray(n), bw)
    bw.close()
}
